qdrant-landing/content/articles/immutable-data-structures.md

```diff
@@ -0,0 +1,236 @@
+A Qdrant-related example of this is [Scalar Quantization](/articles/scalar-quantization/#conversion-to-integers): in order to select proper quantization levels, we have to know the distribution of the data.
+
+{{< figure src="/articles_data/immutable-data-structures/quantization-quantile.png" alt="Scalar Quantization Quantile" caption="Scalar Quantization Quantile" width="70%" >}}
+
+
+Computing this distribution requires knowing all the data in advance, but once we have it, applying scalar quantization is a simple operation.
+
+Let's take a look at a non-exhaustive list of data structures and potential improvements we can get from making them immutable:
+
+|Function| Mutable Data Structure | Immutable Alternative | Potential improvements |
+|----|------|------|------------------------|
+| Read by index | Array | Fixed chunk of memory | Allocate exact amount of memory |
+| Vector Storage | Array or Arrays | Memory-mapped file | Offload data to disk |
+| Read sorted ranges| B-Tree | Sorted Array | Store all data close, avoid cache misses |
+| Read by key | Hash Map | Hash Map with Perfect Hashing | Avoid hash collisions |
+| Get documents by keyword | Inverted Index | Inverted Index with Sorted </br> and BitPacked Postings | Less memory usage, faster search |
+| Vector Search | HNSW graph | HNSW graph with </br> payload-aware connections | Better precision with filters |
+| Tenant Isolation | Vector Storage | Defragmented Vector Storage | Faster access to on-disk data |
+
+
+For more info on payload-aware connections in HNSW, read our [previous article](/articles/filtrable-hnsw/).
+
+This time around, we will focus on the latest additions to Qdrant: 
+- **the immutable hash map with perfect hashing** 
+- **defragmented vector storage**.
+
+### Perfect Hashing
+
+A hash table is one of the most commonly used data structures implemented in almost every programming language, including Rust.
+It provides fast access to elements by key, with an average time complexity of O(1) for read and write operations.
+
+There is, however, the assumption that should be satisfied for the hash table to work efficiently: *hash collisions should not cause too much overhead*.
+
+In regular mutable hash tables, this might be achieved by multiple strategies:
+
+* making the hash table bigger so the probability of collision is lower
+* using a linked list or a tree to store multiple elements with the same hash
+
+However, these strategies have overheads, which become more significant if we consider using high-latency storage like disk.
+
+Indeed, every read operation from disk is order of magnitude slower than reading from RAM, so we want to know the correct location of the data from the first attempt.
+
+In order to achieve this, we can use a so-called perfect hash function (PHF).
+This a special type of hash function is constructed specifically for a given set of keys, and it guarantees no collisions while using minimal memory.
```

Comment:
The RAM usage for the function itself is noticeable (as stated below in the article), so I'd like to replace "minimal memory" with e.g. "minimal amount of buckets".

```suggestion
In order to achieve this, we can use a so-called minimal perfect hash function (MPHF).
This a special type of hash function is constructed specifically for a given set of keys, and it guarantees no collisions while using minimal amount of buckets.
```

---

And if you stick with the word "bucket", we'd better to introduce it beforehand, e.g.
```diff
 There is, however, the assumption that should be satisfied for the hash table to work efficiently: *hash collisions should not cause too much overhead*.
+In a hash table, each key is mapped to a "bucket," a slot where the value is stored.
+When different keys map to the same bucket, a collision occurs.
```